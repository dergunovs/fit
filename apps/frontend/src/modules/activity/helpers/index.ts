import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
