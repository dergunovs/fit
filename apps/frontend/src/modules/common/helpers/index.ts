import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
