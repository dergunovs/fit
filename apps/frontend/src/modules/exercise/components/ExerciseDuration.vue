<template>
  <div :class="$style.timer" :data-active="props.start" data-test="exercise-duration">
    <span>{{ addZero(minutes) }}</span
    >:<span>{{ addZero(seconds) }}</span>
  </div>
</template>

<script setup lang="ts">
import { ref, watch, onMounted } from 'vue';
import { addZero } from 'mhz-helpers';

interface IProps {
  start: boolean;
  stop: boolean;
}

const props = defineProps<IProps>();
const emit = defineEmits<{ stop: [duration: number] }>();

watch(
  () => [props.start, props.stop],
  () => {
    if (props.start) startTimer();
    if (props.stop) stopTimer();
  }
);

const seconds = ref(0);
const minutes = ref(0);

// eslint-disable-next-line no-undef
let interval: NodeJS.Timeout;

function updateTime() {
  seconds.value++;

  if (seconds.value === 60) {
    minutes.value++;
    seconds.value = 0;
  }
}

function startTimer() {
  interval = setInterval(updateTime, 1000);
}

function stopTimer() {
  emit('stop', minutes.value * 60 + seconds.value);
  clearInterval(interval);
}

onMounted(() => {
  if (props.start) startTimer();
});
</script>

<style module lang="scss">
.timer {
  padding: 16px 0;
  font-size: 3rem;
  color: var(--color-gray);
  text-align: center;

  &[data-active='true'] {
    color: var(--color-black);
  }
}
</style>
