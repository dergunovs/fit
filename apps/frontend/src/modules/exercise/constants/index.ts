export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
