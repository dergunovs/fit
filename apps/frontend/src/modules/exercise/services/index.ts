import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
