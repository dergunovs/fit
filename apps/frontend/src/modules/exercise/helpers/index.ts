import {
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';
import { createTempId } from 'mhz-helpers';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

export function getUserEquipmentParams(exercise: IExercise, user?: IUser) {
  const isExerciseHasEquipment = !!exercise.equipment;
  const isExerciseHasEquipmentForWeight = !!exercise.equipmentForWeight?.length;
  const isWeightsRequired = !!exercise.isWeightsRequired;

  const isUserHasEquipment = !!user?.equipments?.some(
    (equipment) => equipment.equipment?._id?.toString() === exercise.equipment?._id?.toString()
  );

  const isUserHasEquipmentForWeight = !!user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some(
      (equipmentForWeight) => equipmentForWeight._id?.toString() === equipment.equipment?._id?.toString()
    )
  );

  return {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  };
}

export function isUserEquipmentMatches(exercise: IExercise, user?: IUser) {
  let result = false;

  const {
    isExerciseHasEquipment,
    isExerciseHasEquipmentForWeight,
    isWeightsRequired,
    isUserHasEquipment,
    isUserHasEquipmentForWeight,
  } = getUserEquipmentParams(exercise, user);

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (!isExerciseHasEquipment && isExerciseHasEquipmentForWeight && isUserHasEquipmentForWeight) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  } else if (
    isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipment &&
    isUserHasEquipmentForWeight
  ) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}`
    : `${exercise.averageDuration.toFixed(1)}`;
}

export function getAverageRepeatsInSet(repeats?: number, sets?: number) {
  return repeats && sets ? Math.round(repeats / sets) : 0;
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise, index) => {
    if (!exercise.dateUpdated || !exercise.duration || !start) return;

    allSteps.push({
      left:
        index === 0
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio,
      right:
        index === exercises.length - 1 && exercises.length !== 1
          ? 0
          : (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    if (!allSteps[index - 1]?.right) return;

    allSteps.push({ left: allSteps[index - 1].right, right: step.left, type: 'rest' });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscle(exercises: IExercise[], title: string, muscle?: string, user?: IUser) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = isUserEquipmentMatches(exercise, user);

    const titleFilter =
      (exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) ||
        exercise.title_en?.toLowerCase().includes(title.toLocaleLowerCase())) &&
      isEquipmentMatches;
    const muscleFilter = exercise.muscles?.some((group) => group._id === muscle) && isEquipmentMatches;

    return muscle ? muscleFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq._id === equipment.equipment?._id)
  );

  if (!equipments?.length) return [];

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?._id === eq._id))
      .map((exToMap) => exToMap._id);

    const availableEquipment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentId) => equipmentId === eq.equipment?._id)
    );

    const options = getWeightsForUserEquipment(availableEquipment);

    return {
      _id: ex._id,
      title: ex.title,
      title_en: ex.title_en,
      options: ex.isWeightsRequired ? options : [0, ...options],
    };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}

export function addSetToExercises(exercises: IExerciseChoosen[]) {
  const set = exercises.slice(-2).map((exercise) => {
    return { ...exercise, _id: createTempId() };
  });

  return [...exercises, ...set];
}

export function updateExercisesIndex(exercises: IExerciseChoosen[], index: number) {
  if (index <= 0 || index >= exercises.length) return exercises;

  const updatedExercises = [...exercises];
  const exerciseToMove = updatedExercises[index];

  updatedExercises.splice(index, 1);
  updatedExercises.splice(index - 1, 0, exerciseToMove);

  return updatedExercises;
}

export function updateExerciseField<T extends keyof IExerciseDone>(
  exercises: IExerciseChoosen[],
  field: T,
  value: IExerciseDone[T],
  id?: string
): IExerciseChoosen[] {
  return exercises.map((exercise) => (exercise._id === id ? { ...exercise, [field]: value } : exercise));
}

export function generateExerciseChoosen(repeats: number, weight?: number, exercise?: IExercise): IExerciseChoosen {
  return {
    _id: createTempId(),
    exercise: {
      _id: exercise?._id || '',
      title: exercise?.title || '',
      title_en: exercise?.title_en || '',
      muscles: exercise?.muscles || [],
      isWeights: exercise?.isWeights || false,
      isWeightsRequired: exercise?.isWeightsRequired || false,
      equipmentForWeight: exercise?.equipmentForWeight,
    },
    repeats,
    weight,
  };
}
