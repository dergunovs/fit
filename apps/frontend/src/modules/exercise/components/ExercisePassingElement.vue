<template>
  <div
    v-if="!props.exercise.isDone"
    :class="$style.element"
    :data-current="props.isCurrentExercise"
    :data-active="isCurrentExerciseActive"
    data-test="exercise-element"
  >
    <div :class="$style.title" :data-current="props.isCurrentExercise" data-test="exercise-title">
      {{
        getExercisePassingTitle(
          props.index,
          props.isCurrentExercise,
          props.exercisesCount,
          props.exercise,
          t('kg'),
          locale
        )
      }}
    </div>

    <template v-if="props.isCurrentExercise">
      <UiButton @click="handleClick" :isDisabled="isButtonDisabled" isTall data-test="exercise-button">
        {{ buttonTitle }}
      </UiButton>

      <UiCheckbox
        v-model="isToFailure"
        :label="t('exercise.doneToFailure')"
        :isDisabled="!isCurrentExerciseActive"
        data-test="exercise-to-failure"
      />

      <ExerciseRepeatsChoice
        v-model="repeats"
        :options="repeatsOptions"
        :title="t('repeat.many')"
        isTall
        data-test="exercise-repeats"
      />

      <ExerciseDurationTimer :start="start" :stop="stop" @stop="sendDurationData" data-test="exercise-duration" />
    </template>
  </div>
</template>

<script setup lang="ts">
import { ref, computed, watch } from 'vue';
import { useI18n } from 'vue-i18n';
import { UiButton, UiCheckbox } from 'mhz-ui';
import { IExerciseDone } from 'fitness-tracker-contracts';

import ExerciseDurationTimer from '@/exercise/components/ExerciseDurationTimer.vue';
import ExerciseRepeatsChoice from '@/exercise/components/ExerciseRepeatsChoice.vue';

import { getExercisePassingTitle } from '@/exercise/helpers';

interface IProps {
  exercise: IExerciseDone;
  activeExerciseId?: string;
  isCurrentExercise: boolean;
  index: number;
  exercisesCount: number;
}

const props = defineProps<IProps>();
const emit = defineEmits<{ start: [id: string]; stop: [exerciseDone: IExerciseDone] }>();

const { t, locale } = useI18n();

const start = ref(false);
const stop = ref(false);

const isToFailure = ref(false);
const repeats = ref(props.exercise.repeats);

const isCurrentExerciseActive = computed(() => props.exercise._id === props.activeExerciseId);

const isButtonDisabled = ref(false);

const repeatsOptions = [
  props.exercise.repeats - 2,
  props.exercise.repeats - 1,
  props.exercise.repeats,
  props.exercise.repeats + 1,
  props.exercise.repeats + 2,
];

const buttonTitle = computed(() => (isCurrentExerciseActive.value ? t('finish') : t('start')));

watch(
  () => props.isCurrentExercise,
  () => {
    if (props.isCurrentExercise) {
      isButtonDisabled.value = true;

      setTimeout(() => {
        isButtonDisabled.value = false;
      }, 500);
    }
  }
);

function handleClick() {
  if (!props.exercise._id) return;

  isButtonDisabled.value = true;

  if (!isCurrentExerciseActive.value) emit('start', props.exercise._id);

  start.value = !isCurrentExerciseActive.value;
  stop.value = isCurrentExerciseActive.value;

  setTimeout(() => {
    isButtonDisabled.value = false;
  }, 500);
}

function sendDurationData(duration: number) {
  emit('stop', { _id: props.exercise._id, duration, isToFailure: isToFailure.value, repeats: repeats.value });
}
</script>

<style module lang="scss">
.element {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 12px;

  &[data-current='true'] {
    padding: 8px 0;
    border-top: 16px solid var(--color-gray);
    border-bottom: 16px solid var(--color-gray);
  }

  &[data-active='true'] {
    border-color: var(--color-transparent);

    &::before,
    &::after {
      position: absolute;
      right: 0;
      left: 0;
      z-index: 1;
      height: 16px;
      pointer-events: none;
      content: '';
      background-image: repeating-linear-gradient(
        -45deg,
        var(--color-success),
        var(--color-success) 16px,
        var(--color-success-dark) 0,
        var(--color-success-dark) 24px
      );
      background-size: 24px 24px;
      animation: stripeMove 1000ms linear infinite;
    }

    &::before {
      top: -16px;
    }

    &::after {
      bottom: -16px;
    }
  }
}

@keyframes stripeMove {
  from {
    background-position: 0 0;
  }

  to {
    background-position: 24px 0;
  }
}

.title {
  display: flex;
  gap: 4px;
  align-items: center;
  color: var(--color-gray-dark-extra);

  &[data-current='true'] {
    font-weight: 700;
    color: var(--color-black);
  }
}
</style>
