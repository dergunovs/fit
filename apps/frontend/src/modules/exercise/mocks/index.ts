import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
