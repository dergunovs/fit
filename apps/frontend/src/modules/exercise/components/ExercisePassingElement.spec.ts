import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import ExercisePassingElement from './ExercisePassingElement.vue';
import ExerciseDuration from './ExerciseDuration.vue';

import { dataTest, wait, wrapperFactory } from '@/common/test';
import { EXERCISE_DONE_FIXTURE, EXERCISE_NOT_DONE_FIXTURE } from '@/exercise/fixtures';
import { getExercisePassingTitle } from '@/exercise/helpers';

const exercise = dataTest('exercise-element');
const title = dataTest('exercise-title');
const button = dataTest('exercise-button');
const toFailure = dataTest('exercise-to-failure');
const repeats = dataTest('exercise-repeats');
const duration = dataTest('exercise-duration');

let wrapper: VueWrapper<InstanceType<typeof ExercisePassingElement>>;

const exerciseNotDone = EXERCISE_NOT_DONE_FIXTURE;
const activeExerciseId = '9999';
const isCurrentExercise = false;
const index = 1;
const exercisesCount = 3;

beforeEach(() => {
  wrapper = wrapperFactory(ExercisePassingElement, {
    exercise: exerciseNotDone,
    activeExerciseId,
    isCurrentExercise,
    index,
    exercisesCount,
  });
});

enableAutoUnmount(afterEach);

describe('ExercisePassingElement', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExercisePassingElement)).toBeTruthy();
  });

  it('shows element only if its not done', async () => {
    expect(wrapper.find(exercise).exists()).toBe(true);

    await wrapper.setProps({ exercise: EXERCISE_DONE_FIXTURE });

    expect(wrapper.find(exercise).exists()).toBe(false);
  });

  it('sets current and active statuses', async () => {
    expect(wrapper.find(exercise).attributes('data-current')).toBe(isCurrentExercise.toString());

    expect(wrapper.find(exercise).attributes('data-active')).toBe(
      (exerciseNotDone._id === activeExerciseId).toString()
    );
  });

  it('shows exercise title', async () => {
    expect(wrapper.find(title).attributes('data-current')).toBe(isCurrentExercise.toString());

    expect(wrapper.find(title).text()).toBe(
      getExercisePassingTitle(index, isCurrentExercise, exercisesCount, exerciseNotDone)
    );
  });

  it('shows exercise controls if its current', async () => {
    expect(wrapper.find(button).exists()).toBe(false);
    expect(wrapper.find(toFailure).exists()).toBe(false);
    expect(wrapper.find(repeats).exists()).toBe(false);
    expect(wrapper.find(duration).exists()).toBe(false);

    await wrapper.setProps({ isCurrentExercise: true });

    expect(wrapper.find(button).exists()).toBe(true);
    expect(wrapper.find(toFailure).exists()).toBe(true);
    expect(wrapper.find(repeats).exists()).toBe(true);
    expect(wrapper.find(duration).exists()).toBe(true);
  });

  it('disables button at start and end to reduce missclick chance', async () => {
    await wrapper.setProps({ isCurrentExercise: true });

    expect(wrapper.find(button).attributes('isdisabled')).toBe('true');

    await wait(501);

    expect(wrapper.find(button).attributes('isdisabled')).toBe('false');

    await wrapper.find(button).trigger('click');

    expect(wrapper.find(button).attributes('isdisabled')).toBe('true');
  });

  it('starts exercise', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('start');

    await wrapper.setProps({ isCurrentExercise: true });

    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.start).toStrictEqual(false);
    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.stop).toStrictEqual(false);

    await wrapper.find(button).trigger('click');

    expect(wrapper.emitted('start')).toHaveLength(1);
    expect(wrapper.emitted()['start'][0]).toStrictEqual([exerciseNotDone._id]);

    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.start).toStrictEqual(true);
    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.stop).toStrictEqual(false);
  });

  it('emits exercise data when stopped', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('stop');

    await wrapper.setProps({ isCurrentExercise: true, activeExerciseId: exerciseNotDone._id });

    await wrapper.find(button).trigger('click');

    await wait(501);

    await wrapper.find(button).trigger('click');

    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.start).toStrictEqual(false);
    expect(wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$props.stop).toStrictEqual(true);

    const durationTime = 12;

    wrapper.findComponent<typeof ExerciseDuration>(duration).vm.$emit('stop', durationTime);

    expect(wrapper.emitted('stop')).toHaveLength(1);
    expect(wrapper.emitted()['stop'][0]).toStrictEqual([
      {
        _id: exerciseNotDone._id,
        duration: durationTime,
        isToFailure: false,
        repeats: exerciseNotDone.repeats,
      },
    ]);
  });

  it('handles exercise to failure checkbox', async () => {
    await wrapper.setProps({ isCurrentExercise: true });

    expect(wrapper.find(toFailure).attributes('isdisabled')).toBe('true');

    await wrapper.setProps({ activeExerciseId: exerciseNotDone._id });

    expect(wrapper.find(toFailure).attributes('isdisabled')).toBe('false');
    expect(wrapper.find(toFailure).attributes('modelvalue')).toBe('false');
  });

  it('handles repeats choice', async () => {
    await wrapper.setProps({ isCurrentExercise: true, activeExerciseId: exerciseNotDone._id });

    expect(wrapper.find(repeats).attributes('options')).toBe(
      [
        exerciseNotDone.repeats - 2,
        exerciseNotDone.repeats - 1,
        exerciseNotDone.repeats,
        exerciseNotDone.repeats + 1,
        exerciseNotDone.repeats + 2,
      ].join()
    );
    expect(wrapper.find(repeats).attributes('modelvalue')).toBe(exerciseNotDone.repeats.toString());
  });
});
