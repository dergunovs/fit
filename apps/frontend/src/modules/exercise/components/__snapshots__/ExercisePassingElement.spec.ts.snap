// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExercisePassingElement > matches snapshot 1`] = `
"<div class="_element_91c235" data-current="false" data-active="false" data-test="exercise-element">
  <div class="_title_91c235" data-current="false" data-test="exercise-title">1. Подтягивание.</div>
  <!--v-if-->
</div>"
`;
